[1, 2단계 - 체스] 폴라(장유진) 미션 제출합니다.

jinchiim opened the pull request:
안녕하세요 아서! 🐹
이번에 리뷰를 부탁드릴 `폴라`라고 합니다!

전반적으로 미션이 정말 어렵기도 하고 시간이 촉박해서 꽤나 진땀 뺀 것 같네요 🥲

최대한 이번 페어와 함께 요구사항 내에서
**가독성이 좋고 한 눈에 체스 게임의 룰을 알 수 있는 코드를 목표**로 작성하게 되었습니다!

궁금한 점이나 고민 거리들은 해당 코드 줄에 코멘트로 달아보겠습니다!
이번 체스 코드 리뷰 잘 부탁드립니다! 😇

jinchiim committed:
docs(README): 1단계 기능 명세서 작성

jinchiim committed:
feat(Row): 체스판의 행 움직임 구현

- 행이 정해진 범위를 벗어나는지 확인하는 테스트 작성

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Column): 체스판의 열 움직임 구현

- 열이 정해진 범위를 벗어나는지 확인하는 테스트 작성

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Coordinate): 입력받은 좌표로 움직이는 메서드 구현

- 입력받은 좌표로 움직이는 지 확인

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Position): 중복되는 로직 조합으로 분리

- 분리 후에도 정상적으로 작동하는지 확인하는 테스트 작성
- 패키지 이동으로 인한 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(PawnStrategy): 폰의 움직임 전략 구현

- 방향에 따른 움직임 전략을 가져오는 메서드 구현
- 움직임이 존재하지 않을 시 에러를 발생하는지 테스트 작성

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Pawn): 갈 수 있는 경로를 가져오는 메서드 구현

- 갈 수 있는 경로를 잘 찾아오는지 테스트
- 첫 움직임이 아닌 경우 2칸을 이동할 수 없는 것을 확인하는 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Pawn): 상수 추가, 변수 이름 변경

- 메서드 수정
- 색깔을 지정하도록 설정

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(PawnAttackStrategy): 폰이 공격할 때 가능한 전략을 구

- 대각선을 갈 수 있도록 구현
- 존재하는 전략이 없다면 에러를 발생하는지 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(PawnStrategy): 색을 확인 후 뒤로 갈 수 없도록 구현

- 뒤로 가는 경우 에러를 발생시키는지 테스트
- validate 메서드 분리

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Position): Position을 빼는 방향 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(KingStrategy): King 의 움직임 구현

- 킹은 한 칸만 이동할 수 있는지 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(QueenStrategy): Queen 의 움직임 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(BishopStrategy): Bishop 의 움직임 구현

- Bishop이 상하좌우로 이동하려고 할 때의 에러 발생 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(RookStrategy): Rook 의 움직임 구현

- Rook이 대각선으로 이동하려고 할 경우 에러 발생 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
style(...): 메서드명 및 예외 메시지 통일

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Bishop): Bishop의 이동 방향을 얻는 메서드 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(King): King의 이동 방향을 얻는 메서드 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Queen): Queen의 이동 방향을 얻는 메서드 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Rook): Rook의 이동 방향을 얻는 메서드 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Knight): Knight의 이동 방향을 얻는 메서드 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Piece): 공통 메서드 인터페이스로 작성

- 같은 색인지 확인하는 메서드 구현
- 공격할 수 있는 Piece 인지 확인

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Position): 입력한 거리만큼 이동하도록 수정

- 입력한 거리만큼 움직이는지 테스트 구현
- 변화에 따른 클래스 내부 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(ChessBoardInitializer): 체스 보드 판을 생성하는 메서드 구현

- 체스 보드 판을 초기화 하는 테스트 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
fix(BishopStrategyTest): 에러 메시지 수정

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(ChessBoard): 체스판에서 체스 말이 움직이는 메서드 구현

- 말이 자신이 갈 위치에 상대 말이 있을 경우 제거하고 그 자리를 차지하는지 테스트
- 빈 칸으로 이동하는지 테스트
- 상대의 말을 이동하려 할 경우 에러 발생 테스트
- 이동 경로에 말이 있는 경우 에러 발생 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(ColumnSymbol): 해당하는 식별자의 위치를 찾는 메서드 구현

- 해당하는 식별자의 위치가 없을 경우 에러를 발생하는 테스트 작성

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Command): 찾는 Command와 식별자가 일치하는지 확인하는 메서드 구현

- 식별자의 일치여부를 가져오는지 확인하는 테스트 작성

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(PieceTranslator): 해당하는 말의 이름을 가져오는 메서드 구현

- 말이 검정말인지 흰말인지에 따른 이름 반환 여부 테스트 작성
- 말이 존재하지 않을 시 이름을 "."으로 반환하는지 테스트 작성

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
fix(PawnAttackStratey): 폰이 대각선에 있는 말만 잡을 수 있도록 수정

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Piece): 말의 색이 검정색인지 확인하는 메서드 인터페이스에 구현

- 인터페이스 내부 구현에 의한 상속하는 객체의 내부 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(InputView): Command를 받아 Split하는 메서드 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(OutputView): 출력 메서드 구현

- 보드판 정보 출력
- 시작 안내 정보 출력

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(ChessGameController): 게임 흐름 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Application): Application 구현

- 의존성 주입

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
style(...): 매직 넘버 상수화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Color): 색상 정보 원시값 상수화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
style(...): 메서드명 수정

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
style(OutputView): 매직 넘버 상수화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(Commands): 코멘드들을 저장하는 Record 생성

- Command 별로 나누어 저장하는지 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(ChessPieceBase): 체스말들의 공통 메서드를 상속받는 추상 클래스 ChessPieceBase 생성

- 생성에 따른 체스 말 객체 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
style(ChessBoard): 패키지 이동

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
style(...): 패키지 이동

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(...): 추상 클래스 생성에 따른 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
delete(RowTest): 패키지 이동에 따른 기존 클래스 삭제

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Command): 일치하는 Command를 찾아오도록 메서드 수정

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Coordinate): 메서드명 변경

- 같은 위치에 존재하는 Row인지 확인하는 메서드 구현
- 메서드 명 변경에 따른 수정 사항 반영

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Position): 메시지를 던지도록 수정

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
style(BishopStrategy): 메서드 명 변경에 따른 수정

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
test(King): King이 자신의 움직임 방향을 가져오는지 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
test(Bishop): Bishop이 자신의 움직임 방향을 가져오는지 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
test(Queen): Queen이 자신의 움직임 방향을 가져오는지 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
test(PawnTest): 메서드 이름 변경

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
feat(CoordinateRequest): Command를 나누어 저장하는 팩토리 메서드 구현

- 적절히 나누어 저장하는지 확인하는 테스트 구현

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
test(QueenStrategyTest): 퀸이 모든 방향을 이동할 수 있는지 테스트

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(ChessGameController): 메서드 분리 및 네이밍 변경

- OutputView 메서드 명 변경
- 메서드명 변경에 따른 Controller 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
test(PawnStrategy): 메서드명 변경에 따른 변화

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
docs(README): 기능 명세서 최종 확인

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Column): 필드 final 추가

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    private void printRow(Map<Coordinate, ChessPiece> board, int row) {
+        for (int column = 0; column < CHESS_BOARD_SIZE; column++) {
+            Coordinate coordinate = new Coordinate(new Row(row), new Column(column));
+            ChessPiece piece = board.get(coordinate);
+            System.out.print(PieceTranslator.getName(piece));
+        }
```
## 상황

이번 체스 미션에서 말들의 위치와 정보를 담는 자료구조로 해당하는 좌표의 말들을 찾기 편한
`Map`을 선택하게 되었습니다.

그래서 도메인 내에서는 board 라는 이름으로 잘 사용했는데요,
이를 출력하기 위해서는 마찬가지로 해당 정보는 `View`에게 넘겨주어야만 했습니다.

## 고민

사실 `Map<Coordinate, ChessPiece>`를 가진다 하더라도,
`View`에서 해당하는 좌표에 말이 존재하는지 보기만 하기 때문에 문제가 없다고 생각하고 있기는 합니다.

## 질문

Map이야 `UnModifiableMap`으로 전달해준다는 점에서 괜찮다고 하더라도,
`Map`에서 `Coordinate`를 이용해 찾기 위해 객체를 만들어야 한다는 점이 굉장히 맘에 걸렸습니다.  😢

만약 아서라면 이에 대해서 어떻게 생각하시는지 궁금합니다!

Hyunta replied:
제 생각에는 지금 구조에서 가장 직관적으로 해결해주신 것 같습니다👍🏻
개발 내용을 떠나서, 일단 돌아가도록 만들고 리팩터링을 고민하신 작업 흐름이 매우 적절해보입니다💯
개인적으로 복잡한 애플리케이션을 만들 때는 생각나는데로 먼저 구현하고 리팩터링하는 것이 더 효율적인 것 같아요.
앞으로도 고민이 되는 곳이 있다면 일단 만들고 `//TODO` 주석을 적절히 이용하시면 좋을 것 같습니다!

view에서 domain 로직을 호출하는 것이 마음에 걸리셨다면,
해볼 수 있는 선택지는 3가지 정도 떠오르네요.
1. Map 구현체 중 순서를 유지하는 LinkedHashMap을 사용한다.
2. Controller에서 출력 순서를 갖고있는 List<Coordinate>를 같이 전달한다.
3. Coordinate가 Comparable 인터페이스를 구현하도록 만들어서 Coordinate를 정렬해서 출력한다.

위 3가지 방법 중에 괜찮다고 생각하시는 방법을 적용해보시는 것도 좋을 것 같습니다!
더 나은 아이디어가 생기셨다면 그걸 적용하셔도 좋습니다☺️

jinchiim replied:
첫번째 선택지가 `Blank`라는 객체를 만들면거 사용하기에 어울리는 것 같아,
`LinkedHashMap`으로 변동하여 사용했습니다!

jinchiim commented on src/main/java/view/dto/CoordinateRequest.java:
```diff
@@ -0,0 +1,16 @@
+
+import view.util.ColumnSymbol;
+
+public record CoordinateRequest(int row, int column) {
+
+    private static final int CHESS_BOARD_SIZE = 8;
+
+    public static CoordinateRequest fromCommand(String command) {
+        String[] split = command.split("");
+        int column = ColumnSymbol.from(split[0]).getPosition();
+        int row = CHESS_BOARD_SIZE - Integer.parseInt(split[1]);
+
+        return new CoordinateRequest(row, column);
```
## 상황

입력 중 `move a1 a2`와 같은 식으로 받을 때
이를 `ChessBoard`에 전달해주기 위해 `split()` 후 `List<>.get(0)` 과 같은 식으로
이전에는 전달해주고 있었는데요. 

## 고민

그러다 보니, 가독성이 좋은 코드를 작성하고 싶다는 시선에서 바라보았을때
이 **`List<>`가 무엇인지, 이 안에 무엇이 담겨있는 것인지 알 수가 없어**
가독성이 떨어진다고 느꼈습니다.

그래서 Dto를 만들어 적절한 네이밍으로 전달해주고자 했는데요.

이후 `Controller`의 가독성은 올라갔지만 이와 같이 `Dto` 내부에서 체스 보드 크기만큼 받은 값을 빼주는,
비즈니스 로직이라고 여겨지는 과정을 가진 팩토리 메서드가 추가되어졌습니다.

## 질문

이에 대해서 페어와 오랜 시간 얘기를 나누었지만,
가독성이 너무 떨어진다 생각해 이와 같은 결정을 내리게 되었는데요.

아서는 Dto에서 팩토리 메서드로 이런 로직을 가지는 것에 대해
어떻게 생각하는지 궁금합니다! 🥰

Hyunta replied:
DTO를 생성하는데 어떤 command가 넘어오는지 모르는 것은 좋은 설계는 아니라고 생각됩니다.
질문을 듣고 어떤 값이 command로 넘어오는지 확인하기 위해서, 연관된 객체를 살펴봐야했습니다.
적절한 생성 로직은 아닌 것 같아요.

그런데 이 로직을 DTO에서 하는게 아니라 Coordinator에서 해줘야할 것 같아요.
해당 DTO를 도메인을 변경해주는 시점이 앞당겨져야된다고 생각합니다. 
그렇기 때문에 해당 DTO 자체가 정말 필요한가를 고민해봐야할 것 같아요.
Commands (DTO) -> CoordinateRequest -> Coordinate 로 2번에 걸쳐서 변환되는 것 보다는
Commands(DTO) -> Coordinate 로 한번에 변환 시켜주는게 코드가 간결해질 것 같습니다.
제가 Commands(DTO) 쪽에 남긴 리뷰를 참고해서 한번 고민해보세요!

jinchiim replied:
말씀하신 피드백 대로, String 값을 받아 Coordinate로 변경하도록 수정하고 보니,
`Dto`의 필요성을 못 느껴 DTO를 삭제하고 `Command`에서 특정 커멘드가 맞는지 확인하는 메서드를 추가했습니다!

jinchiim committed:
style(KnightStrategy): 컨벤션 통일

Co-authored-by: pricelees <priceelees@gmail.com>

Hyunta commented on src/main/java/view/dto/Commands.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.List;
+import view.util.Command;
+
+public record Commands(List<String> commands) {
```
DTO끼리는 네이밍 컨벤션을 맞춰보면 어떨까요?
Command는 domain인데, Commands가 dto로 분류되어있어서 헷갈립니다.

jinchiim replied:
해당 Dto는 피드백을 반영하며 삭제했습니다!

Hyunta commented on src/main/java/controller/ChessGameController.java:
```diff
@@ -0,0 +1,92 @@
+    private ChessBoard initializeBoard() {
+        Commands commands = inputView.receiveCommands();
+
+        if (Command.START != commands.command()) {
```
Commands는 DTO인데 비즈니스 로직에 계속 사용되네요,
DTO를 Command enum으로 변환해서 로직에 사용하면 어떨까요?

jinchiim replied:
Command Enum으로 변환하고 또 다시 확인하는 반복 작업을 현재 하고 있어,
`Command` 내부에 특정 커멘드인지 확인하는 메서드를 두었습니다! 😁

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,83 @@
+
+    private static final int ROW_DIRECTION = 0;
+    private static final int COLUMN_DIRECTION = 1;
+    private static final Map<Coordinate, ChessPiece> board = ChessBoardInitializer.createInitialBoard();
```
board를 static으로 관리하게되면 모든 ChessBoard가 같은 board를 가지게 되겠네요.
static으로 선언하는 것과 인스턴스 변수로 선언하는 것은 어떤 차이가 있을까요?
https://stackoverflow.com/questions/7026507/why-are-static-variables-considered-evil%22

jinchiim replied:
테스트를 작성하면서도 느끼긴 했지만,
테스트를 하기도 어렵고 다른 객체로 부터 변화가 생겨 문제가 생긴다면,
 원인을 알아채기 어려울 것 같아요 🧐

말씀해주신 피드백 대로 인스턴스 변수로 선언하여 사용했습니다!

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,83 @@
+        return destinationPiece != null && !destinationPiece.isSameColor(current);
+    }
+
+    public Map<Coordinate, ChessPiece> getBoard() {
+        return Collections.unmodifiableMap(board);
+    }
```
UnmodifiableMap👍🏻

Hyunta commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,37 @@
+
+public class OutputView {
+
+    private static final int CHESS_BOARD_SIZE = 8;
```
OutputView에서 꼭 알고있을 필요는 없을 것 같습니다.
Map의 크기에 따라 출력해주도록 해도 view의 역할은 충분히 할 수 있지 않을까요?

jinchiim replied:
`Linked` 와 `Blank`라는 객체를 생성하면서
 Map의 크기에 따라 출력하도록 변경했습니다!

Hyunta replied:
private static final int ROW_SIZE = 8;
그런데 여전히 상수로 8이라는 값을 알고있네요,
이 값이 꼭 필요한 값인가요?

jinchiim replied:
해당 부분에서 말씀해주신 부분이,

- `OutputView`에서 `8`이라는 상수를 반드시 가져야 하는가?
-  `8`이라는 정보 없이`Map<Coordinate, Piece>` 로부터 충분히 보드 모양을 만들 수 있지 않은가?

두가지가 있다고 생각했는데요. 현재는 전자라고 생각해
이미 `ColumnSymbol` 에서 총 몇개의 `Row`혹은 `Column`이 존재하는지 알 수 있을 것 같아
이곳에서 불러오도록 수정했으나

만약 후자의 의미에서 질문을 주셨다면, 제가 부족한지
`LinkedHashMap`만으로 어떤 기준으로 보드 모양을 만들 수 있을지 
생각해봤지만 좀 어려운 것 같습니다ㅜ 😢

후자라면 혹시 어떤 키워드를 주실 수 있을지 여쭤보고 싶습니다!

Hyunta commented on src/main/java/domain/coordinate/Coordinate.java:
```diff
@@ -0,0 +1,54 @@
+public class Coordinate {
+
+    private final Row row;
+    private final Column column;
```
위치를 나타내기 위한 적절한 객체를 만들어주셨네요👍🏻
패키지의 위치가 Row, Column이랑 같이 있어도 될 것 같은데 별도로 분리해주신 이유가 있을까요?

jinchiim replied:
처음에는 `Row` 와 `Column` 이 단지 Coordinate에서 쓰기 보다는 
방향을 나타내던 `List<Integer>`에서 사용할 것 같아 분리하게 되었습니다!

다만 시간 부족으로... `List<Integer>`가 그대로 남게 되면서 그 잔해 라고 볼 수 있을 것 같아요

변경한 현재도 `Coordinate`에서만 사용해 패키지 이동 했습니다!

Hyunta commented on src/main/java/controller/ChessGameController.java:
```diff
@@ -0,0 +1,92 @@
+            throw new IllegalArgumentException("이미 시작한 상태 입니다.");
+        }
+        return true;
+    }
```
이것도 Command Enum 안으로 넣을 수 있을 것 같습니다.

jinchiim replied:
Command Enum 안에 일치하는지 확인하는 메서드를 이동시켰습니다!

Hyunta commented on src/main/java/controller/ChessGameController.java:
```diff
@@ -0,0 +1,92 @@
+
+    private Coordinate createCoordinate(CoordinateRequest coordinateRequest) {
+        return new Coordinate(new Row(coordinateRequest.row()), new Column(coordinateRequest.column()));
+    }
```
String을 전달해서 Coordinate를 만들 수 있도록 하면 어떨까요?
A1 -> Coordinate(Row: A, Column:1)
그럼 Controller에서 굳이 생성하지 않고 Coordinate 클래스를 이용할 수 있을 것 같습니다

jinchiim replied:
피드백을 주신 바와 같이 `coordinate`에서 `a1`과 같은 값을 받을 수 있도록 변경했습니다!

변경하면서 전체적으로 보드판 시야(?)를 바꾸게 되었습니다.
이전에는 a0을 입력하면 블랙 룩을 선택하게 되었는데요, 현재는 실제 보드판과 같게 `1~8` 그리고 `a~h` 로
`a1`이 흰색 룩을 선택하도록 변경했습니다!

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,83 @@
+        validateBeforePlay(destination, currentTurn, piece);
+
+        List<Integer> direction = piece.getDirection(start, destination,
+                canAttack(currentTurn, destination));
```
List<Integer> direction은 일급컬렉션으로 감싸질 필요가 있어보입니다.
해당 클래스 상위에 정의된 상수도 옮길 수 있겠네요.

jinchiim replied:
해당 부분을 `Record`를 사용하여 감싸게 되었는데요.
이에 대해 아서의 의견이 궁금해 코멘트 달아보겠습니다! 🥰

Hyunta replied:
저는 도메인 객체를 Record로 감싸는 것을 사용하지는 않습니다.
class를 정의해서 필요한 메서드를 작성해주는 것이 가독성도 좋을 뿐더러 예기치 않은 문제를 발생시키지 않아서요.
예시로 getter의 경우 convention으로 getXXX() 형태를 가져가야하는데 record에서는 XXX() 형태로 만들어주기 때문에 특정 라이브러리를 사용하기 위해서는 getXXX()를 정의해야합니다.
그리고 record의 탄생 배경 자체가 불변 데이터를 전달하는 캐리어의 목적으로 설계되었습니다. 그래서 도메인에서는 사용하지 않습니다.
https://openjdk.org/jeps/395

jinchiim replied:
Class로 정의해서 바꿔보겠습니다! 
`record`에 대해서 더 공부해봐야겠네요..
자료와 의견 감사합니다!

jinchiim replied:
`record` 테코톡도 들었으면서 `dto`같은 모습에 `record`를 도메인에 써버렸네요 🥲
클래스로 변경했습니다!

자료와 의견 감사합니다!

Hyunta commented on src/test/java/domain/chessboard/ChessBoardTest.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Order(1)
+    @DisplayName("상대 말을 제거한 뒤 이동한다.")
+    @Test
```
나중에 ChessBoardTest를 진행할 때 순서에 영향을 받겠네요.
Order 말고 다른 방법으로 구성할 수 없을까요?

jinchiim replied:
순서에 영향받지 않는 테스트 작성을 위해 인스턴스 변수로 옮기면서 
`Order`을 지웠습니다!

Hyunta commented on src/main/java/view/util/Command.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Arrays;
+
+public enum Command {
```
util 패키지로 분류하는 기준이 있을까요?
util 패키지로 관리하면 어떤 장단점이 있나요?

jinchiim replied:
개인적으로 해당 부분들은 전역적으로 사용할 수 있기도 하고,
각각 하나의 패키지로 묶어주기에는 오히려 패키지가 많아져서 헷갈릴수도 있겠다고 생각했어요!

그래서 3가지를 전부 묶어줄 적절한 이름을 찾다가 어디서든 활용이 가능하단 의미에서 `util`이라는 이름을 붙이게 되었는데요.

다만 피드백을 받고 관련하여 찾아보던 도중 `util`이라는 패키지는 보통 **도메인과 연관성이 적고 
정말 말 그래도 어디서든 가져다 사용할 수 있을 만큼 단순한 구성**이 좋다는 정보를 보게 되었습니다.

확실히 현재의 `Util` 패키지에 들어가있는 3가지는 도메인과 연관성이 적지도, 구성이 단순하지도 않은 것 같아요.

그래서 `symbol` 과 `PieceTranslator` 의 경우 특정 문구를 특정 무언가로 바꾸어준다는 점에서 `translator` 라는
패키지에 넣고 `Command`의 경우 
`command` 보다 적절한 이름이 없어 같은 이름의 패키지에 넣어주었습니다! 😊

Hyunta replied:
개인적으로 util 클래스는 최후의 보루로 사용합니다.
만약 조금이라도 연관된 클래스가 보인다면 해당 클래스와 관련된 패키지 내부에 만들거나, 클래스에 정의하는게 바람직하다고 생각해요.
폴라가 말해준 것 처럼 util은 말 그대로 정말 범용적으로 쓰이는 것에 한정되서 사용합니다.
예를 들어 시간과 관련된 유틸, 문자열 포매팅과 관련된 유틸 등 도메인과 전혀 관계 없이 공용으로 사용하는 툴을 정의할 때만 사용합니다.
앞으로도 미션을 진행하면서 util 없이 개발하는 것을 목표로 최대한 객체에게 책임을 위임하는 연습을 해보세요.

Hyunta commented on src/main/java/domain/piece/strategy/pawn/PawnAttackStrategy.java:
```diff
@@ -0,0 +1,43 @@
+    final int rowDifference;
+    final int columnDifference;
+    final List<Integer> direction;
+    final Color allowedPawnColor;
```
모두 private이어도 괜찮을 것 같네요

jinchiim replied:
앗.. 작성하다가 놓쳤나 보네요 😞
모두 private 으로 변경했습니다!

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,83 @@
+    private boolean canAttack(Color current, Coordinate destination) {
+        ChessPiece destinationPiece = board.get(destination);
+
+        return destinationPiece != null && !destinationPiece.isSameColor(current);
```
[질문]
만약 빈자리에 null이 아니라 ChessPiece를 상속받는 빈 말이 있었다면 null에 대해 처리안해도 됐을 것 같아요.
반영하라는건 아닙니다, null로 위치시키는 것과 `빈 말` 객체로 위치시켜놓는 것 각각 장단점이 무엇일까요?

jinchiim replied:
제가 생각했을 때 null 로 위치했을 때의 장점은 직관적이라는 것 한가지 인 것 같습니다.
한 눈에 봐도 비어있다는 것을 알 수 있으니깐요.

하지만 null 이라는 것은 결국 **정말 내가 의도한 대로 비어있는 값인건지, 실수로 null이 들어가지게 된 것인지 판단하기 어렵다** 라는 단점이 생길 것 같아요.

판단하기 어렵다는 것은 에러가 났을 때 원인을 파악하기도 어려운 것으로 직결된다고도 생각합니다. 😕

빈 말이 있는 것이 좋은 것 같아 `Blank` 라는 빈 말 객체를 만들고,
`NONE` 이라는 컬러를 만들어 변경하였습니다!

Hyunta replied:
빈말을 정의하는 것과 null로 두는 것은 각각 장단점이 있는 것 같아요.
앞으로 3,4단계 미션을 진행하시면서 느껴보세요!

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,83 @@
+    private boolean canAttack(Color current, Coordinate destination) {
+        ChessPiece destinationPiece = board.get(destination);
+
+        return destinationPiece != null && !destinationPiece.isSameColor(current);
```
[제안]
1. 메서드 이름
개인적으로 메서드 이름이 canAttack이 아니라 isAttack으로 가야하지 않을까 싶어요
지금 움직이는 위치가 공격 위치라는 것을 말해주는 메서드인 것 같아서요.

2. !destinationPiece.isSameColor(current)
이 로직은 체스에서는 중요한 로직인 것 같아요. 매번 `!`로 부정문으로 나타내기 보다는 적군인지 확인하는 메서드를 만들면 어떨까요?

jinchiim replied:
1. 판단은 한번 뿐이니 `isAttack` 도 어울리는 이름 같아 변경했습니다!

2. 개인적으로는 `isNot~` 과 같이 `Not`이 들어가는게 오히려 가독성이 떨어진다고 생각하는 의견이 있어
    검증은 모두 긍정문으로 만들려고 노력했었는데,
    사실상 한 곳을 제외하고는 전부 `!`라는 부정문을 사용하고 있어 변경하는게 더 어울리는 것 같기도 하고
    메서드 이름을 변경하면 좋을 것 같아
    해당 메서드를 `isOpponentColor`로 변경했습니다!

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,83 @@
+
+            hasPieceOnPath(coordinate, destination);
+        }
+    }
```
메서드 이름은 validate인데 실제로 말을 움직이면서 이동이 가능한지 확인하고 있어요.
메서드 이름을 적절하게 변경할 수 있을까요?

그리고 해당 메서드에서 start를 움직이는 바람에 처음에 방어적 복사를 하고 시작하게 됐어요.
start(coordinate) 를 움직이는게 아니라 iteration을 돌면서 direction의 배수를 올리면서 검증하는 방식을 사용해볼 수 있지 않을까요? row 나 column이 보드 사이즈를 벗어나기 전까지 늘려보는 방식으로요

jinchiim replied:
해당 메서드의 이름이 말씀하신 대로 이동을 하며 가능한지 확인하고 있어
말을 옮기는 것과 같이 묶어 `move`라는 이름으로 변경했습니다!

> start(coordinate) 를 움직이는게 아니라 iteration을 돌면서 direction의 배수를 올리면서 검증하는 방식을 사용해볼 수 있지 않을까요? row 나 column이 보드 사이즈를 벗어나기 전까지 늘려보는 방식으로요

이 부분은 당시 페어와 논의 했었는데요!
`direction`의 배수를 올리면서 검증하기 위해 계속해서 `Coordinate`라는 인스턴스를 만들어야 했기에
불필요한 메모리와 작업이 많이 든다고 생각했습니다. 

두번째로는 보드의 끝이 목표가 아닌, 정해진 좌표가 목표이기 때문에
보드 사이즈를 벗어나기 전까지 늘리는 방식이 결국은 정해진 좌표가 될때까지 늘리는 방식과 일치하다고 생각합니다!

이에 대해서 아서는 어떻게 생각하실까요? 의견을 여쭤보고 싶습니다! 🧐

Hyunta replied:
제 의도는 늘리는 대상이 start가 아니라 direction이어도 되지 않냐는 말이었어요.
기물이 움직일 수 있는지를 확인할 떄 start에 direction을 계속 더해가면서 판단을 하고 있는데,
기물의 위치는 가만히 냅두고 iteration을 돌면서 index를 1에서 계속 1씩 증가를 시키고
`start + (direction * index)` 를 하게 되면 start에 조작이 안이뤄지니 더 괜찮지 않을까 싶었습니다.

지금은 방어적 복사를 잘 해주셔서 로직상 문제는 전혀 없겠지만,
개인적으로 인자로 받은 객체를 수정하는 일을 최대한 안만드는 편이 유지보수하기 쉽다고 생각합니다.

jinchiim replied:
> 개인적으로 인자로 받은 객체를 수정하는 일을 최대한 안만드는 편이 유지보수하기 쉽다고 생각합니다.

그렇네요.. 인스턴스가 생기는게 싫어 유지보수를 버리는 선택을 한 것 같아요.

말씀해주신 방법을 적용하다가,
`direction` 을 외부에서 `1`씩 돌리면 매직 넘버가 생길 것 같아

`Coordinate` 와 내부의 방어적 복사를 삭제하고,

while문을 도는 동안`Coordinate.next(Direction direction)`메서드 호출로
 내부에서 주어진 방향에 맞게 1칸 이동한 좌표를 주도록 변경했습니다!

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+
+    private ChessPiece findPiece(Coordinate start) {
+        return board.get(start);
+    }
```
```suggestion
    private ChessPiece findPiece(Coordinate coordinate) {
        return board.get(coordinate);
    }
```

Hyunta commented on src/main/java/domain/piece/Pawn.java:
```diff
@@ -0,0 +1,66 @@
+        return pawnPath.getDirection();
+    }
+
+    private void validate(PawnStrategy pawnStrategy, Coordinate coordinate, int rowDifference) {
+        validateCanMovePawn(coordinate, rowDifference);
+        pawnStrategy.validateMoveBackward(color);
+    }
+
+    private void validateAttack(PawnAttackStrategy pawnStrategy, Coordinate coordinate, int rowDifference) {
+        validateCanMovePawn(coordinate, rowDifference);
+        pawnStrategy.validateMoveBackward(color);
+    }
+
+    private void validateCanMovePawn(Coordinate coordinate, int rowDifference) {
+        if (!isFirstPosition(coordinate) && Math.abs(rowDifference) == MAX_PAWN_DISTANCE_DIFFERENCE) {
+            throw new IllegalArgumentException("폰은 처음에만 2칸을 이동할 수 있습니다.");
+        }
+    }
```
외부에서 호출하는 것은 getDirection인데,
실제로는 움직일 수 있는지 검증까지 하고 있네요.

실제로 움직일 수 있는지 검증하는 책임은 Board로 옮기는게 어떨까요?

jinchiim replied:
개인적으로는 **폰이 처음 이동일 시 2칸을 이동할 수 있다** 는 것은
폰의 이동 전략에 포함이 된다고 생각했습니다.
그래서 오히려 보드판에서 `폰은 처음에만 2칸을 이동할 수 있다`는 메시지를 던지는 것이
더 어색하게 느껴졌던 것 같아요.

이에 대해 아서의 의견은 어떠실지 궁금합니다! 😁

Hyunta replied:
한칸 움직이나, 두칸 움직이나 같은 방향을 가져가는 구조로 이해했습니다.
그런데 getDirection을 호출했는데 2칸 이동한다는 메세지가 나오는게 어색하지 않나요?

메서드 이름을 수정해서 움직일 수 있는 방향과 거리를 다 계산한다는 식으로 바꾸던지,
아니면 방향에 대한 검증, 거리에 대한 검증을 각각 메서드로 분리하던지 명확한 메서드 이름을 주는 것은 필요해보입니다.

질문해주신 폰은 처음 이동할 때 2칸 움직일 수 있다는 저 또한 PawnStrategy에 있어야 한다는 점에는 동의해요

jinchiim replied:
> 그런데 getDirection을 호출했는데 2칸 이동한다는 메세지가 나오는게 어색하지 않나요?

확실히 `get`에서 해당 에러를 뱉어내니 어색하게 느껴지네요 🤔
두가지 모두 고려해서 수정해보겠습니다!

jinchiim replied:
> 그런데 getDirection을 호출했는데 2칸 이동한다는 메세지가 나오는게 어색하지 않나요?

외부에서 보았을때는 `getDirection`에서 해당 에러가 나오는군요... 🤔

폰의 첫번째 위치인지 확인하는 메서드는 내부에 그대로 두고,
`2칸` 이동이 가능한 폰인지는 board에서 확인하도록 이동하고 메서드 네이밍을 `validateCanMoveTwoDistance`로 변경했습니다!

Hyunta reviewed (changes requested):
안녕하세요 폴라,
체스 미션 1,2단계 잘 진행해주셨네요.
도메인 로직이 복잡한데 흐름을 깔끔하게 잡아주셔서 읽기 좋았습니다.
관련해서 몇가지 커멘트 남겨놨으니 확인 부탁드려요.
궁금한점이 있으면 DM이나 커멘트로 남겨주세요!

jinchiim committed:
refactor(ChessBoardInitializer): 체스판을 바라보는 시야 변경

- Black이 0,0이 아닌 8,1이 되도록 초기화 변경
- 체스판 좌표 변경에 따른 전략 변경
- 아무것도 존재하지 않는 Blank 체스판 생성

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(Command): Coordinate 생성자 Private으로 제어

- Command에서 스스로 확인하도록 변경
- 생성자 변경에 따른 변경사항들 적용

Co-authored-by: pricelees <priceelees@gmail.com>

jinchiim committed:
refactor(ChessBoard): static이 아닌 인스턴스 변수로 선언하도록 변화

- Order() 삭제

jinchiim committed:
refactor(ChessBoardGameController): 게임의 턴을 Board에서 관리하도록 변경

jinchiim committed:
chore(PawnStrategy): 사용하지 않는 println() 삭제

jinchiim committed:
refactor(PawnStrategy): private 접근 제어자 추가

jinchiim committed:
refactor(ChessBoard): isAttack으로 이름 변경

- 부정문 긍정이 되도록 변경

jinchiim committed:
refactor(Blank): Color를 default로 넣도록 변경

jinchiim committed:
refactor(ChessBoardInitializer): 초기 보드를 Blank를 추가해 세팅하는 메서드 추가

jinchiim committed:
refactor(Color): 빈 색 추가

jinchiim committed:
refactor(PieceTranslator): Blank.class 추가

jinchiim committed:
refactor(board): board 에서 remove가 아닌 put으로 변경

jinchiim committed:
refactor(board): board 에서 put이 아닌, replace로 변경

jinchiim committed:
refactor(ChessBoardInitializer): LinkedHashMap으로 바꾼 후 initialize 방법 변경

jinchiim committed:
refactor(OutputView): Coordinate를 만드는게 아닌, values를 8개씩 잘라 출력하도록 변경

jinchiim committed:
refactor(Direction): Direction 불변 객체 생성

- 불변 객체 생성에 따른 변화

jinchiim committed:
refactor(...): Destination 생성에 따른 테스트 수정

jinchiim committed:
refactor(Coordinate): 사용하지 않는 생성자 삭제

jinchiim committed:
refactor(Column): 패키지 이동

jinchiim committed:
refactor(Command): 움직임 상태인지 확인하는 과정 Command 내부로 이동

jinchiim committed:
style(...): 패키지 이름 수

jinchiim committed:
refactor(ChessPiece): 반대 색인지 확인하는 메서드 이름 변경

jinchiim committed:
refactor(ChessBoard): 메서드 이름 변경

jinchiim commented on src/main/java/domain/direction/Direction.java:
```diff
@@ -0,0 +1,13 @@
+package domain.direction;
+
+import java.util.List;
+
+public record Direction(int rowDirection, int columnDirection) {
+
+    private static final int ROW_DIRECTION = 0;
+    private static final int COLUMN_DIRECTION = 1;
+
+    public static Direction from(List<Integer> direction) {
+        return new Direction(direction.get(ROW_DIRECTION), direction.get(COLUMN_DIRECTION));
+    }
```
## 상황

처음에는 `Coordinate`를 이용하려고 했으나,
사실 방향을 뜻하는 의미에서 좌표라는 이름의 객체를 사용하면 안될 것 같아,

int rowDirection, int columnDirection 을 가지는 `Direction` 이라는 클래스를 새로 만들려고 했습니다.

## 고민

그러나 막상 만들고 보니
전부 결국 그냥 `get`과 같이 **값만 전달하는 형태**로 이루어져 있더라구요.
가지고 있는 것은 결국 방향뿐이니 `Direction`에서 할만한 일도 없다고 느껴졌어요.

결국은 **`Dto`와 같은 행동만을 한다**고 느껴 
`record`로 만들어 값만 전달하도록 작성하게 되었습니다.

## 고민

이 `Direction`은 사실 좌표를 움직일때 말고는 사용하지 않고 있고,
값만 전달하기 때문에 이와 같은 형태로 둬도 된다고 생각이 들면서도

`Direction`이 언젠가 다른 일을 맡을 수 있을 정도로
**아주 완전히 도메인과 떨어진 역할을 맡은 것은 아닌 것 같아** `record`를 사용하는게 맞았을까 고민이 듭니다 😞

이게 `record`를 `dto`로써 자주 사용해서 거부감이 드는 것인지 잘 모르겠어요.
그러나 아무리 생각해도 굳이 현 상태에서 클래스로 만들어줄 이유는 찾지 못해 이대로 두게 되었습니다.

이에 대해 아서는 어떻게 생각하실까요? 🤔

Hyunta replied:
위에서 record와 관련하여 답변을 드렸는데,
해당 내용도 참고하시면 좋을 것 같습니다ㅎㅎ

jinchiim commented:
안녕하세요 아서!
전반적으로 체스판의 시야도 바꾸고 구조도 바꾸면서 많은 시간이 걸려 리뷰 요청을 드리게 되었네요
리뷰를 확인했나 고민도 들었을 것 같아요. 죄송합니다  🥺

변경사항에 대한 내용은
피드백을 적어주신 코멘트에 답글로 달아두는 것이 편하실 것 같아 코멘트에 달아두었습니다!

이번 리뷰도 잘 부탁드립니다!

Hyunta commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,17 @@
+import view.InputView;
+import view.OutputView;
+
+public class Application {
```
게임상 오류가 있네요.
```
start
move a2 a3
move g8 g6
```
하면 나이트가 움직이지 않습니다.

jinchiim replied:
앗 방금 확인해보니 리팩토링을 하면서 상수에서 놓친 부분이 있었네요 🥲
`position`의 크기를 0에서 1로, 7에서 8로 바꾸었습니다!

Hyunta commented on src/main/java/domain/piece/strategy/pawn/PawnStrategy.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Arrays;
+import java.util.List;
+
+public enum PawnStrategy {
+
+    UP(1, 0, List.of(1, 0), Color.WHITE),
+    DOWN(-1, 0, List.of(-1, 0), Color.BLACK),
+    UP_UP(2, 0, List.of(1, 0), Color.WHITE),
+    DOWN_DOWN(-2, 0, List.of(-1, 0), Color.BLACK);
```
[질문]
strategy를 enum으로 구성하면서 UP, DOWN 처럼 중복되는 Enum 값들이 생기게 됐습니다.
혹시 이런 부분들을 Direction에 묶어놓고 Strategy에서 받아서 쓰는 형태에 대해서는 어떻게 생각하시나요?

jinchiim replied:
현재 `Strategy`에 중복이 많아서 제거가 가능할 것 같아요!

사실 시간이 부족해서 포기했던 부분인데, 이번에 적용해보겠습니다!

`Direction`이라는 `Enum`에 방향을 명시한 후, 
각 객체에게 맞는 `Direction`을 생성할때 `default`로 생성하도록 변경해주었습니다!

Hyunta reviewed (changes requested):
안녕하세요 폴라,
구조가 많이 변경되어서 제가 놓친 부분이 생길 수 있을 것 같네요.
혹시 만약 제가 이전 구조와 착각해서 남긴 커멘트가 있으면 알려주세요.
몇가지 커멘트를 남겨놨는데 3,4단계 가기전에 반영해서 가면 좋을 것 같아서 재요청보냅니다.
궁금한 점이 있으면 DM 이나 커멘트로 남겨주세요.

jinchiim committed:
refactor(Position): 리팩토링 하며 놓쳤던 Position 최소, 최대값 수정

jinchiim committed:
refactor(Direction): 클래스로 변경

jinchiim committed:
refactor(ChessBoard): 인자를 그대로 두고, 반복하며 다음 위치를 갈 수 있는지 확인하도록 수정

- Position 에서 해당 direction 만큼 이동하는 메서드 네이밍 수정
- 사용하지 않는 메서드 copied 삭제

jinchiim committed:
refactor(ChessBoard): 폰이 2칸을 갈 수 있는지 확인하는 메서드 ChessBoard로 이동

jinchiim committed:
refactor(OutputView): 상수 제거

jinchiim committed:
refactor(Pawn): 사용하지 않는 파라미터 삭제

jinchiim committed:
refactor(ChessBoard): 메서드 이름 재 변경

jinchiim reviewed (commented):
안녕하세요 아서 😄 

이번에 전체적으로 피드백을 반영하면서 
여러가지 변화가 생겨서 설명을 드리기 위해 코멘트를 남깁니다!

### 1.  `Direction` 관리

기존에는 각 체스 말마다 `Direction`을 관리하고 있었는데요.
반복되는 부분들이 존재하고, 객체가 너무 많아져 하나의 `Direction` 이라는 `Enum` 에서
관리하도록 변경하였습니다!

```java
    public static final Set<Direction> STRAIGHT_DIRECTION = Set.of(UP, DOWN, LEFT, RIGHT);
    public static final Set<Direction> DIAGONAL_DIRECTION = Set.of(UP_RIGHT_DIAGONAL, UP_LEFT_DIAGONAL,
            DOWN_LEFT_DIAGONAL, DOWN_RIGHT_DIAGONAL);
    public static final Set<Direction> KNIGHT_DIRECTION = Set.of(UP_UP_RIGHT_DIAGONAL, UP_UP_LEFT_DIAGONAL,
            LEFT_LEFT_UP_DIAGONAL, LEFT_LEFT_DOWN_DIAGONAL, RIGHT_RIGHT_UP_DIAGONAL, 
....
```

### 2. `Strategy` 추상화

이전과 달리 `Direction`을 다른 곳에서 관리하면서 
`Strategy`가 각 하나씩 있을 필요가 사라졌습니다.

그래서 `Strategy`를 상속받은 추상클래스 `PieceStrategy`를 생성하고
그 안에서 각 객체의 `Direction` 만 주입하도록 변경하게 되었습니다.

```java
public class PieceStrategy implements Strategy {

    private final Set<Direction> directions;

    public PieceStrategy(Set<Direction> directions) {
        this.directions = directions;
    }

    @Override
    public Direction findDirection(int rowDifference, int columnDifference) {
        return directions.stream()
                .filter(direction -> direction.isEqualDirection(rowDifference, columnDifference))
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException("존재하는 움직임 전략이 없습니다."));
    }
}
```

### 3. `Pawn` 색깔 별로 분리

`Pawn`의 색이 두가지나 되면서, 
두가지의 상수와 두가지의 조건문이 생기게 되었었는데요.

시작하는 위치와 `Direction` 만 다르기 때문에 `ChessPieceBase`를 `extend`한 `Pawn` 이라는 추상 클래스를 만들어
동일한 작동을 한곳에서 관리하고
두가지 색으로 `Pawn`을 나누어 각 `Direction`을 따로 관리하게 되었습니다.

```java

public abstract class Pawn extends ChessPieceBase {

    private final PieceStrategy pieceStrategy;

    public Pawn(Color color, Set<Direction> directions) {
        super(color);
        this.pieceStrategy = new PieceStrategy(directions);
    }

    @Override
    public Direction getDirection(Coordinate start, Coordinate destination, boolean canAttack) {
        int rowDifference = start.calculateRowDifference(destination);
        int columnDifference = start.calculateColumnDifference(destination);
        validateDifference(rowDifference, columnDifference);

        if (canAttack) {
            pieceStrategy.findDirection(rowDifference, columnDifference);
        }
        return pieceStrategy.findDirection(divideValueByAbs(rowDifference), divideValueByAbs(columnDifference));
    }
```

큰 부분들은 이 정도이고,
작은 변화들과 궁금증은 코멘트로 달아놓았습니다!

이번 리뷰도 잘 부탁드립니다! 😊

jinchiim committed:
feat(Strategy): 전략에서 하는 공통 행동을 선언

jinchiim committed:
feat(PieceStrategy): 움직임 전략 중, 일치하는 전략을 찾는 메서드 정의

- 존재하는 전략이 없는 경우 테스트 구현

jinchiim committed:
feat(Pawn): 폰의 움직임을 추상화한 클래스 선언

jinchiim committed:
feat(...Pawn): 색 별로, Pawn을 상속받아 default 색과 directions 선언

- 초기 위치와 같은 곳인지 확인하는 메서드 선언
- 해당하는 폰의 방향을 가져오는지 테스트 구현
- start 위치가 폰의 시작 위치와 일치하는지 확인하는 테스트 구현

jinchiim committed:
refactor(Command): 사용하지 않는 메서드 삭제

jinchiim committed:
refactor(Direction): 잘못된 방향 설정 수정

jinchiim committed:
test(Blank): 빈 말판 테스트 작성

jinchiim committed:
refactor(...): Strategy 추상화 및 Direction 객체 생성에 따른 수정

- Knight 테스트 추가

jinchiim committed:
refactor(Command): Command 패키지 이동

Hyunta commented:
>해당 부분에서 말씀해주신 부분이,
OutputView에서 8이라는 상수를 반드시 가져야 하는가?
8이라는 정보 없이Map<Coordinate, Piece> 로부터 충분히 보드 모양을 만들 수 있지 않은가?
두가지가 있다고 생각했는데요. 현재는 전자라고 생각해
이미 ColumnSymbol 에서 총 몇개의 Row혹은 Column이 존재하는지 알 수 있을 것 같아
이곳에서 불러오도록 수정했으나

지금 수정해주신 방향도 충분히 괜찮은 것 같습니다.
전달드렸던 의도는 전자와 후자가 합쳐진 내용이었습니다.
주어진 정보로 출력할 수 있을 것 같은데 왜 8이라는 숫자가 꼭 필요해보이지 않아서요.

제가 구상했던 내용은 컨트롤러에서 보드의 순서에 맞춰서 피스를 가져오는 것이에요.
Row와 Column 정보를 출력해야하는 순서대로 맞춰서 Board의 Piece를 `List<List<String>>` 혹은 DTO 형태로 받아놓습니다.
그리고 OutputView로 전달하면 OutputView에서는 한줄에 하나씩 출력만 해주는거에요.

그런데 지금 구성해주신 방법도 충분히 괜찮은 방법이라고 생각됩니다.

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,113 @@
+        Direction direction = piece.getDirection(start, destination, isAttack);
+        validateNoPieceOnPath(start, destination, direction);
+
+        if (!isAttack && piece instanceof Pawn) {
```
instanceof 말고 메서드로 해결해볼 수는 없을까요?
아래 글을 한번 읽어보고 고민해보세요
https://tecoble.techcourse.co.kr/post/2021-04-26-instanceof/

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,113 @@
+    private void validateDestination(Coordinate destination, Color currentTurn) {
+        ChessPiece chessPiece = findPiece(destination);
+
+        if (chessPiece.getClass() == Blank.class) {
```
마찬가지로 Class를 직접 비교하지 않고 메서드를 만들 수 있을 것 같습니다

Hyunta commented on src/main/java/domain/chessboard/ChessBoard.java:
```diff
@@ -0,0 +1,113 @@
+            return;
+        }
+
+        if (!findPiece(coordinate).getClass().equals(Blank.class)) {
```
여기도 비슷한 로직이 있네요

Hyunta reviewed (approved):
안녕하세요 폴라,
고민을 많이하시는 것이 코드에서 느껴지네요, 설계를 잘 변화해 나가고 있습니다.
커멘트 반영해주신 내용들 확인했고 추가로 클래스 비교 관련해서 커멘트 남겨놨습니다.
다음 단계 진행하시면서 반영해주세요!
궁금한 사항이 있으시면 DM 주시거나 커멘트를 남겨주세요!
